File: javascript/after-refactor/PromotionTextDecorator.js
class PromotionTextDecorator {
  constructor(product) {
    this.product = product;
    this.promoTextLines = [];
  }

  static decorate(product) {
    return new PromotionTextDecorator(product).decorate();
  }

  decorate() {
    if (this.product.promoted) {
      this.includePromotionText();
    } else {
      this.includeDescription();
    }
    this.updatePromotionText();
    return this.product.promotionText;
  }

  updatePromotionText() {
    this.product.promotionText = this.promoTextLines.join('\n');
  }

  includeDescription() {
    this.promoTextLines.push(this.product.description);
  }

  includePromotionText() {
    this.includeDiscountText();
    this.includeFreeShippingText();
    this.includeDescription();
    this.includeFreeShippingText();
    this.includeDiscountText();
  }

  includeDiscountText() {
    const discountDollar = this.getDiscountInDollar();
    if (this.product.hideDiscount || discountDollar <= 0) {
      return;
    }
    const discountText = this.getDiscountText(discountDollar);
    this.promoTextLines.push(discountText);
  }

  getDiscountText(discount) {
    return `Don't miss the deal, $${discount} off!`;
  }

  getDiscountInDollar() {
    return (this.product.oldCentPrice - this.product.centPrice) / 100;
  }

  includeFreeShippingText() {
    if (!this.product.freeShipping) {
      return;
    }
    const freeShippingText = this.getFreeShippingText();
    this.promoTextLines.push(freeShippingText);
  }

  getFreeShippingText() {
    return 'Free shipping!';
  }
}

module.exports = {
  PromotionTextDecorator,
};

File: java/src/main/java/promotiontext/after/PromotionTextDecorator.java
package promotiontext.after;

import java.util.ArrayList;
import java.util.List;
import promotiontext.model.Product;

/**
 * PromotionTextDecorator generates promotion text for a product.
 */
public class PromotionTextDecorator {
  private Product product;
  private List<String> promoTextLines;

  protected PromotionTextDecorator(Product product) {
    this.product = product;
    promoTextLines = new ArrayList<>();
  }

  /**
   * Generate and set promotion text for a product.
   *
   * @param product the product
   * @return the generated promotion text
   */
  public static String decorate(Product product) {
    return new PromotionTextDecorator(product).decorate();
  }

  private String decorate() {
    if (product.isPromoted()) {
      includePromotionText();
    } else {
      includeDescription();
    }
    updatePromotionText();
    return product.getPromotionText();
  }

  private void updatePromotionText() {
    product.setPromotionText(String.join("\n", promoTextLines));;
  }

  private void includeDescription() {
    promoTextLines.add(product.getDescription());
  }

  private void includePromotionText() {
    includeDiscountText();
    includeFreeShippingText();
    includeDescription();
    includeFreeShippingText();
    includeDiscountText();
  }

  private void includeDiscountText() {
    int discountDollar = getDiscountInDollar();
    if (product.isHideDiscount() || discountDollar <= 0) {
      return;
    }
    String discountText = getDiscountText(discountDollar);
    promoTextLines.add(discountText);
  }

  private String getDiscountText(int discountDollar) {
    return String.format("Don't miss the deal, $%d off!", discountDollar);
  }

  private int getDiscountInDollar() {
    return (product.getOldCentPrice() - product.getCentPrice()) / 100;
  }

  private void includeFreeShippingText() {
    if (!product.isFreeShipping()) {
      return;
    }
    String freeShippingText = getFreeShippingText();
    promoTextLines.add(freeShippingText);
  }

  private String getFreeShippingText() {
    return "Free shipping!";
  }
}
